staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -214,17 +215,49 @@ func newETCD3Check(c storagebackend.Config, timeout time.Duration, stopCh <-chan
 		}
 		ctx, cancel := context.WithTimeout(context.Background(), timeout)
 		defer cancel()
-		// See https://github.com/etcd-io/etcd/blob/c57f8b3af865d1b531b979889c602ba14377420e/etcdctl/ctlv3/command/ep_command.go#L118
 		now := time.Now()
-		_, err := client.Get(ctx, path.Join("/", c.Prefix, "health"))
-		if err != nil {
-			err = fmt.Errorf("error getting data from etcd: %w", err)
-		}
+		err := prober.Probe(ctx)
 		lastError.Store(err, now)
 		return err
 	}, nil
 }
 
+func newETCD3Prober(c storagebackend.Config) (*etcd3Prober, error) {
+	client, err := newETCD3Client(c.Transport)
+	if err != nil {
+		return nil, err
+	}
+	return &etcd3Prober{
+		Client: client,
+		prefix: c.Prefix,
+	}, nil
+}
+
+type etcd3Prober struct {
+	*clientv3.Client
+	closed atomic.Bool
+	prefix string
+}
+
+func (p *etcd3Prober) Close() error {
+	if p.closed.CompareAndSwap(false, true) {
+		return p.Client.Close()
+	}
+	return fmt.Errorf("prober was closed")
+}
+
+func (p *etcd3Prober) Probe(ctx context.Context) error {
+	if p.closed.Load() {
+		return fmt.Errorf("prober was closed")
+	}
```

Comment:
This isn't thread-safe.

The fact that you check `closed` here doesn't mean that Close() will not be called in the meantime (after checking it above but before calling p.Client below.

The only way to making it really thread-safe is to use locks and lock (reader-lock) this whole function.

Reply:
Right. missed that race.